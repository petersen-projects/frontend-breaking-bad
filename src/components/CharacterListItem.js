import React from 'react';

export const CharacterListItem = ({
   name, age, img, status
}) => (
   <div className="list-row">
      {/* <div className="list-row__image-container">
      <img src={img}></img>
      </div> */}
      <div className="list-row__info">
         <div className="list-row__title">{name}</div>
         <div>Age: {age}</div>
         <div>Status: {status}</div>
      </div>
   </div>
);

export default CharacterListItem;