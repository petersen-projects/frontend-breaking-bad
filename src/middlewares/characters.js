import { calculateAgeFromBirthdayString } from '../utils/functions';

export const parseCharactersResultFromWs = (characters) => {
   return characters.map(character => ({
      id: character.char_id,
      name: character.name,
      birthday: character.birthday,
      age: calculateAgeFromBirthdayString(character.birthday),
      img: character.img,
      nickname: character.nickname,
      status: character.status
   }));
};