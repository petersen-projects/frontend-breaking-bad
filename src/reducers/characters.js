import {
   GET_CHARACTERS_REQUEST_BEGIN,
   GET_CHARACTERS_REQUEST_ERROR,
   GET_CHARACTERS_REQUEST_SUCCESS,

   SET_CHARACTERS_NUMBER_OF_PAGES,
   SET_CHARACTERS_OFFSET,
   SET_CHARACTERS_DATA,
   SET_CHARACTERS_FILTERED_DATA,

   RESET_CHARACTERS_OFFSET,
   RESET_CHARACTERS_FILTERED_DATA
} from '../actions/characters';

const charactersReducerDefaultState = {
   data: [],
   filteredData: [],
   fetching: false,
   offset: 0,
   numberOfPages: 0,
   limit: 10
};

export default (state = charactersReducerDefaultState, action) => {
   switch(action.type) {
      case GET_CHARACTERS_REQUEST_BEGIN:
         return {
            ...state,
            fetching: true
         };
      case GET_CHARACTERS_REQUEST_SUCCESS:
         return {
            ...state,
            data: [...state.data, ...action.data],
            fetching: false
         };
      case GET_CHARACTERS_REQUEST_ERROR:
         return {
            ...state,
            data: [],
            fetching: false
         };
      case SET_CHARACTERS_OFFSET:
         return {
            ...state,
            offset: action.offset
         };
      case SET_CHARACTERS_NUMBER_OF_PAGES:
         return {
            ...state,
            numberOfPages: action.numberOfPages
         };
      case SET_CHARACTERS_DATA:
         return {
            ...state,
            data: action.data
         };
      case SET_CHARACTERS_FILTERED_DATA:
         return {
            ...state,
            filteredData: [...state.filteredData, ...action.filteredData]
         }
      case RESET_CHARACTERS_OFFSET:
         return {
            ...state,
            offset: 0
         }
      case RESET_CHARACTERS_FILTERED_DATA:
         return {
            ...state,
            filteredData: []
         }
      default:
         return state;
   }
};
