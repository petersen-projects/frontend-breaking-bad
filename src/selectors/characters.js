export const selectCharacters = (
   {data, name = '', status = '', offset = 0, limit = 10}
) => {
   const filteredData = filterData({data, name, status});
   return filteredData.slice(offset * limit, (offset * limit) + limit);
}

export const getNumberOfPages = ({data, name = '', status = '', limit = 10}) => {
   return Math.ceil(
      filterData({data, name, status}).length / 10
   );
}

const filterData = ({data, name, status}) => {
   return data.filter((character) => {
      const textMatch = character.name.toLowerCase().includes(name.toLowerCase());
      const statusMatch = status ? character.status.toLowerCase().includes(status) : true;

      return textMatch && statusMatch;
   });
}