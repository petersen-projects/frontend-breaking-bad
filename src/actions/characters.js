import axios from 'axios';
import { parseCharactersResultFromWs } from '../middlewares/characters';

export const GET_CHARACTERS_REQUEST_BEGIN = 'GET_CHARACTERS_REQUEST_BEGIN';
export const GET_CHARACTERS_REQUEST_ERROR = 'GET_CHARACTERS_REQUEST_ERROR';
export const GET_CHARACTERS_REQUEST_SUCCESS = 'GET_CHARACTERS_REQUEST_SUCCESS';

export const SET_CHARACTERS_DATA = 'SET_CHARACTERS_DATA';
export const SET_CHARACTERS_OFFSET = 'SET_CHARACTERS_OFFSET';
export const SET_CHARACTERS_NUMBER_OF_PAGES = 'SET_CHARACTERS_NUMBER_OF_PAGES';
export const SET_CHARACTERS_FILTERED_DATA = 'SET_CHARACTERS_FILTERED_DATA';

export const RESET_CHARACTERS_OFFSET = 'RESET_CHARACTERS_OFFSET';
export const RESET_CHARACTERS_FILTERED_DATA = 'RESET_CHARACTERS_FILTERED_DATA';

const WEBSERVICE_CHARACTERS = 'https://www.breakingbadapi.com/api/characters';

export const startGetAllCharactersFromWs = () => {
   return (dispatch) => {
      const getCharactersFromWs =
         axios.get(WEBSERVICE_CHARACTERS);      
      dispatch({ type: GET_CHARACTERS_REQUEST_BEGIN });

      return getCharactersFromWs.then(({data}) => {
         dispatch(getCharactersSuccess(data));
      });
   }
}

export const getCharactersError = (error) => ({
   type: GET_CHARACTERS_REQUEST_ERROR,
   error
});

export const getCharactersSuccess = (data) => ({
   type: GET_CHARACTERS_REQUEST_SUCCESS,
   data: parseCharactersResultFromWs(data)
});

export const setCharactersOffset = (offset) => ({
   type: SET_CHARACTERS_OFFSET,
   offset
});

export const resetCharactersOffset = () => ({
   type: RESET_CHARACTERS_OFFSET
});

export const setCharactersNumberOfPages = (numberOfPages) => ({
   type: SET_CHARACTERS_NUMBER_OF_PAGES,
   numberOfPages
});

export const setCharactersData = (data) => ({
   type: SET_CHARACTERS_DATA,
   data
});

export const setCharactersFilteredData = (filteredData) => ({
   type: SET_CHARACTERS_FILTERED_DATA,
   filteredData
});

export const resetCharactersFilteredData = () => ({
   type: RESET_CHARACTERS_FILTERED_DATA
});

export const startGetCharactersFromWs = (
   {limit = 10, offset = 1, name = '', status = ''},
   shouldCleanData = false
) => {
   return (dispatch) => {
      const getCharactersFromWs =
         axios.get(normalizeCharactersWebserviceUrl({
            limit, offset, name, status
         }));

      dispatch({ type: GET_CHARACTERS_REQUEST_BEGIN });

      return getCharactersFromWs.then(({data}) => {
         if (!shouldCleanData) {
            dispatch(getCharactersSuccess(data));
            dispatch(setCharactersHasMoreData(data.length === limit));
         } else {
            dispatch(setCharactersData([]));
            dispatch(setCharactersOffset(0));
            dispatch(getCharactersSuccess(data));
         }
      }).catch((error) => dispatch(getCharactersError(error)));
   }
};

const normalizeCharactersWebserviceUrl = ({limit = 10, offset = 1, name = '', status = ''}) => {
   const limitQuery = `?limit=${limit}`;
   const offsetQuery = `&offset=${offset > 1 ? offset * 10 : offset}`;
   const nameQuery = name !== '' ? `&name=${name.replaceAll(' ', '+')}` : '';
   const statusQuery = status !== '' ? `&status=${status}` : '';

   return `${WEBSERVICE_CHARACTERS}${limitQuery}${offsetQuery}${nameQuery}${statusQuery}`;
};
