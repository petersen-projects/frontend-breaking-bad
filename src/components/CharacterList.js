import React, {
   useRef,
   useCallback,
   useEffect
} from 'react';
import { useSelector, useDispatch } from 'react-redux';

import {
   setCharactersOffset,
   resetCharactersOffset,
   startGetAllCharactersFromWs,
   setCharactersFilteredData,
   resetCharactersFilteredData,
   setCharactersNumberOfPages
} from '../actions/characters';
import { selectCharacters, getNumberOfPages } from '../selectors/characters';
import CharacterListItem from './CharacterListItem';
import Loading from './Loading';

const CharacterList = () => {
   const dispatch = useDispatch()

   const data = useSelector(state => state.characters.data)
   const filteredData = useSelector(state => state.characters.filteredData)
   const fetching = useSelector(state => state.characters.fetching)
   const numberOfPages = useSelector(state => state.characters.numberOfPages)
   const offset = useSelector(state => state.characters.offset)
   const limit = useSelector(state => state.characters.limit)
   const name = useSelector(state => state.filters.name)
   const status = useSelector(state => state.filters.status)

   useEffect(() => {
      dispatch(startGetAllCharactersFromWs())
   }, [])

   useEffect(() => {
      updateFilteredData(data)
   }, [data])

   useEffect(() => {
      if (offset > 0) {
         resetOffset()
      } else {
         refreshFilteredData()   
      }
   }, [name, status])

   useEffect(() => {
      if (offset > 0) {
         paginateData()
      } else {
         refreshFilteredData()
      }
   }, [offset])

   const observer = useRef()
   const lastCharacterRef = useCallback(node => {
      if (Math.ceil(filteredData.length / limit) >= numberOfPages) return
      if (observer.current) observer.current.disconnect()

      observer.current = new IntersectionObserver(entries => {
         if (entries[0].isIntersecting) {
            dispatch(setCharactersOffset(offset + 1))
         }
      })

      if (node) observer.current.observe(node)
   })

   const updateFilteredData = (() => {
      if (data.length) {
         dispatch(
            setCharactersFilteredData(
               selectCharacters({data, limit})
            )
         )
         dispatch(
            setCharactersNumberOfPages(
               getNumberOfPages({data, limit})
            )
         )
      }
   })

   const resetOffset = (() => {
      dispatch(resetCharactersOffset())
   })

   const refreshFilteredData = (() => {
      dispatch(
         resetCharactersFilteredData()
      )
      dispatch(
         setCharactersFilteredData(
            selectCharacters({data, limit, offset, name, status})
         )
      )
      dispatch(
         setCharactersNumberOfPages(
            getNumberOfPages({data, limit, offset, name, status})
         )
      )
   })

   const paginateData = () => {
      dispatch(
         setCharactersFilteredData(
            selectCharacters({
               data, offset, name, status
            })
         )
      )
   }

   return (
      <div className="list-body">
         { fetching == true ? <Loading /> : null}
         {
            filteredData.length === 0 ? (
               <div className="list-body__no-data-message">No data available</div>
            ) : (
               filteredData.map((character, index) => {
                  if (filteredData.length === index + 1) {
                     return (
                        <div ref={lastCharacterRef} key={index}>
                           <CharacterListItem {...character} />
                        </div>
                     )
                  } else {
                     return <CharacterListItem key={index} {...character} />
                  }
                     
               })
            )
         }
      </div>
   )
}

export default CharacterList;
