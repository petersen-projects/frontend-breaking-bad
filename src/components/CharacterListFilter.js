import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { setNameFilter, setCharacterStatusFilter } from '../actions/filters';
import { startGetAllCharactersFromWs } from '../actions/characters';

export const CharacterListFilter = () => {
   const name = useSelector(state => state.filters.name)
   const status = useSelector(state => state.filters.status)
   const dispatch = useDispatch()

   const onNameChange = (e) => {
      dispatch(setNameFilter(e.target.value))  
   }
   const onStatusFilterChange = (e) => {
      dispatch(setCharacterStatusFilter(e.target.value))
   }

   return (
      <div>
         <input
            type="text"
            placeholder="Search by character name..."
            value={name}
            onChange={onNameChange}
         />
         <select
            className=""
            value={status}
            onChange={onStatusFilterChange}
         >
            <option value="">Choose one Status</option>
            <option value="alive">Alive</option>
            <option value="deceased">Deceased</option>
         </select>
         <button>Search</button>
      </div>
   )
}

export default CharacterListFilter;