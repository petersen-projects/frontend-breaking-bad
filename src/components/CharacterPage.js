import React from 'react';

import CharacterList from './CharacterList';
import CharacterListFilter from './CharacterListFilter';
import Header from './Header';

export const CharacterPage = () => (
   <div className="character-page-container">
      <Header />
      <CharacterListFilter />
      <CharacterList />
   </div>
);

export default CharacterPage;