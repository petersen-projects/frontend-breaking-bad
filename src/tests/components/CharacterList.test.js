import React from 'react';
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { Provider } from 'react-redux';
import { shallow, mount } from 'enzyme';

import { CharacterList } from '../../components/CharacterList';
import characters from '../fixtures/characters';

const mockStore = configureMockStore([thunk]);
const initialState = {
   characters: {
      data: [],
      filteredData: []
   },
   filters: {}
};

test('should render CharactersList with characters', () => {
   const wrapper = shallow(
      <Provider store={mockStore(initialState)}>
         <CharacterList filteredData={characters} />
      </Provider>
   );
   expect(wrapper).toMatchSnapshot();
});

test('should render CharactersList with empty message', () => {
   const wrapper = mount(
      <Provider store={mockStore(initialState)}>
         <CharacterList filteredData={[]} />
      </Provider>
   );

   expect(wrapper).toMatchSnapshot();
   expect(wrapper.find('.message').text()).toBe("no data :(");
});
