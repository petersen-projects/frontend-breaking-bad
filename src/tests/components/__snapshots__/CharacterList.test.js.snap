// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CharactersList with characters 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <CharacterList
    filteredData={
      Array [
        Object {
          "birthday": "09-07-1958",
          "id": 1,
          "img": "https://images.amcnetworks.com/amc.com/wp-content/uploads/2015/04/cast_bb_700x1000_walter-white-lg.jpg",
          "name": "Walter White",
          "status": "Presumed dead",
        },
        Object {
          "birthday": "09-24-1984",
          "id": 2,
          "img": "https://vignette.wikia.nocookie.net/breakingbad/images/9/95/JesseS5.jpg/revision/latest?cb=20120620012441",
          "name": "Jesse Pinkman",
          "status": "Alive",
        },
        Object {
          "birthday": "08-11-1970",
          "id": 3,
          "img": "https://s-i.huffpost.com/gen/1317262/images/o-ANNA-GUNN-facebook.jpg",
          "name": "Skyler White",
          "status": "Alive",
        },
        Object {
          "birthday": "07-08-1993",
          "id": 4,
          "img": "https://media1.popsugar-assets.com/files/thumbor/WeLUSvbAMS_GL4iELYAUzu7Bpv0/fit-in/1024x1024/filters:format_auto-!!-:strip_icc-!!-/2018/01/12/910/n/1922283/fb758e62b5daf3c9_TCDBRBA_EC011/i/RJ-Mitte-Walter-White-Jr.jpg",
          "name": "Walter White Jr.",
          "status": "Alive",
        },
        Object {
          "birthday": "Unknown",
          "id": 7,
          "img": "https://images.amcnetworks.com/amc.com/wp-content/uploads/2015/04/cast_bb_700x1000_mike-ehrmantraut-lg.jpg",
          "name": "Mike Ehrmantraut",
          "status": "Deceased",
        },
      ]
    }
  />
</ContextProvider>
`;

exports[`should render CharactersList with empty message 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <CharacterList
    filteredData={Array []}
  >
    <div
      className="list-body"
    >
      <div
        className="message"
      >
        no data :(
      </div>
    </div>
  </CharacterList>
</Provider>
`;
